mixer/pkg/cache/lruCache.go

```diff
@@ -0,0 +1,272 @@
+// index of the last entry in the list, while the sentinel's next field holds the index of
+// the next entry in the list.
+//
+// The LRU algorithm we currently use is classically simple. Both getting or setting
+// a cache entry puts that entry at the head of the LRU list. When we need to make
+// room in the cache, we always just plop the current tail from the list.
+//
+// Once this code has been in active use for a while in a real system, we
+// should evaluate whether fancier LRU regimes would improve overall perf.
+// For example, we could merely bump entries up in the list by one slot per
+// use instead of putting them at the head of the list. Or we could require
+// N accesses before promoting an entry to the head. Etc.
+//
+// If we do get more sophisticated, we could consider introducing an explicit MRU section
+// of the list (MRU=Most Recently Used). Entries in that portion of the list aren't shuffled
+// around upon use, they are treated as a whole distinct class. The idea is that for frequently
+// used entries, we can avoid messing up processor caches by moving these entries.
+//
+// Due to the use of the time.Time.UnixNano function in this code, expiration
+// will fail after the year 2262. Sorry, you'll need to upgrade to a newer version
+// of Istio at that time :-)
+
+type lruCache struct {
+	sync.Mutex
+	entries           []lruEntry
+	sentinel          *lruEntry
+	lookup            map[interface{}]int // keys => entry index
+	stats             Stats
+	defaultExpiration time.Duration
+	evictionTicker    *time.Ticker
+	baseTime          int64 // nanoseconds
+}
+
+// lruEntry is used to hold a value in the ordered lru list represented by the entry slice
+type lruEntry struct {
+	next       int         // index of next entry
+	prev       int         // index of previous entry
+	key        interface{} // cache key associated with this entry
+	value      interface{} // cache value associated with this entry
+	expiration int64       // nanoseconds
+}
+
+// entry 0 in the slice is the sentinel node
+const sentinelIndex = 0
+
+// NewLRUCache creates a new cache with an LRU and time-based eviction model.
+//
+// Cache eviction is done on a periodic basis. Individual cache entries are evicted
+// after their expiration time has passed. The periodic nature of eviction means that
+// cache entries tend to survive around (expirationTime + (evictionInterval / 2))
+//
+// In addition, when the cache is full, adding a new item will displace the item that has
+// been referenced least recently.
+//
+// defaultExpiration specifies the default minimum amount of time a cached
+// entry remains in the cache before eviction. This value is used with the
+// Set function. Explicit per-entry expiration times can be set with the
+// SetWithExpiration function instead.
+//
+// evictionInterval specifies the frequency at which eviction activities take
+// place. This should be >= to 1 second.
```

Comment:
curious/thinking out loud: at what point does it become more efficient to only do expiry calculation at `Get` time, or, say, when capacity crosses some threshold (like 80% of `maxEntries`)?  Is this related to ratio of `Get` calls/second to `evictionInterval` somehow?

Reply:
Having Get check expiration time on every call makes Get around 40% slower. Even if you do expiration checks in Get, you still need a reaper running in the background to reclaim dead unused entries for the sake of memory use.

I'm definitely not happy about having the eviction logic do a full sequential scan over all cache entries. That's a potential large amount of churn in the processor cache + TLB which just slows everything down. I originally started with multiple time-based buckets so that eviction would run on one bucket at a time and would systematically evict everything in the bucket. Alas, that clashed with my desire to have a single pointer-less slice for all the entries.

Do you think the slop in expiration time is problematic?